native/cocos/engine/Engine.cpp

```diff
@@ -309,89 +293,45 @@ void Engine::tick() {
 }
 
 void Engine::doRestart() {
-    cc::EventDispatcher::dispatchRestartVM();
+    cc::event::broadcast<events::RestartVM>();
     destroy();
     CC_CURRENT_APPLICATION()->init();
 }
 
-bool Engine::handleEvent(const OSEvent &ev) {
-    bool isHandled = false;
-    OSEventType type = ev.eventType();
-    if (type == OSEventType::TOUCH_OSEVENT) {
-        cc::EventDispatcher::dispatchTouchEvent(OSEvent::castEvent<TouchEvent>(ev));
-        isHandled = true;
-    } else if (type == OSEventType::MOUSE_OSEVENT) {
-        cc::EventDispatcher::dispatchMouseEvent(OSEvent::castEvent<MouseEvent>(ev));
-        isHandled = true;
-    } else if (type == OSEventType::KEYBOARD_OSEVENT) {
-        cc::EventDispatcher::dispatchKeyboardEvent(OSEvent::castEvent<KeyboardEvent>(ev));
-        isHandled = true;
-    } else if (type == OSEventType::CONTROLLER_OSEVENT) {
-        cc::EventDispatcher::dispatchControllerEvent(OSEvent::castEvent<ControllerEvent>(ev));
-        isHandled = true;
-    } else if (type == OSEventType::CUSTOM_OSEVENT) {
-        cc::EventDispatcher::dispatchCustomEvent(OSEvent::castEvent<CustomEvent>(ev));
-        isHandled = true;
-    } else if (type == OSEventType::WINDOW_OSEVENT) {
-        isHandled = dispatchWindowEvent(OSEvent::castEvent<WindowEvent>(ev));
-    } else if (type == OSEventType::DEVICE_OSEVENT) {
-        isHandled = dispatchDeviceEvent(OSEvent::castEvent<DeviceEvent>(ev));
-    }
-    isHandled = dispatchEventToApp(type, ev);
-    return isHandled;
-}
-
-bool Engine::handleTouchEvent(const TouchEvent &ev) { // NOLINT(readability-convert-member-functions-to-static)
-    cc::EventDispatcher::dispatchTouchEvent(ev);
-    return dispatchEventToApp(OSEventType::TOUCH_OSEVENT, ev);
-}
-
 Engine::SchedulerPtr Engine::getScheduler() const {
     return _scheduler;
 }
 
-bool Engine::dispatchDeviceEvent(const DeviceEvent &ev) { // NOLINT(readability-convert-member-functions-to-static)
-    bool isHandled = false;
-    if (ev.type == DeviceEvent::Type::MEMORY) {
-        cc::EventDispatcher::dispatchMemoryWarningEvent();
-        isHandled = true;
-    } else if (ev.type == DeviceEvent::Type::ORIENTATION) {
-        cc::EventDispatcher::dispatchOrientationChangeEvent(ev.args[0].intVal);
-        isHandled = true;
-    }
-    return isHandled;
-}
-
-bool Engine::dispatchWindowEvent(const WindowEvent &ev) {
+bool Engine::redirectWindowEvent(const WindowEvent &ev) {
     bool isHandled = false;
     if (ev.type == WindowEvent::Type::SHOW ||
         ev.type == WindowEvent::Type::RESTORED) {
-        emit(static_cast<int>(ON_RESUME));
+        static_cast<BaseEngine*>(this)->emit<EngineStatusChange>(ON_RESUME);
```

Comment:
Why need to convert to a base class when engine inherits from BaseEngine and should be able to call it directly?

Reply:
It is a template function that is not inherited.